### PR TITLE
Sardinian Keyboard

### PR DESCRIPTION
**Pull request by adrmzz**
With corpus.txt (but empty database).

### CODE REVIEW

**Review by adrmzz — COMMENTED** [review]
updated keyboards with reorganization of special characters (inspired by the [Catalan update](https://github.com/ubports/keyboard-component/pull/124)) and addition of pronouns (·dd, ·l, ·nche, ·nde, ·mi, ·ti, ·si, ·nos, ·bos)

File: debian/control [hunk 1]
```diff
@@ -119,7 +119,8 @@ Depends: autopilot-qt5 (>= 1.4),
          ubuntu-keyboard-portuguese (>= ${source:Version}),
          ubuntu-keyboard-romanian (>= ${source:Version}),
          ubuntu-keyboard-russian (>= ${source:Version}),
-         ubuntu-keyboard-scottish-gaelic (>= ${source:Version}),
+         ubuntu-keyboard-sardinian (>= ${source:Version}),
+	 ubuntu-keyboard-scottish-gaelic (>= ${source:Version}),
```
Comment (Flohack74) [inline debian/control]: Please fix indentation here

**Review by Flohack74 — DISMISSED** [review]

**Review by UniversalSuperBox — CHANGES_REQUESTED** [review]
With the DB populated in this PR, I believe it will actually cause the keyboard to overflow the available space on the system-image on some devices. We may need to merge this keyboard without the corpus database, disabling autocorrect.

The corpus text submitted here has no license information. Please add this information to the `debian/copyright` file so we can ensure license compatibility.

**Comment by adrmzz** [comment]
I am stuck here. Can anyone lend a hand? Could we simply re-use [AnySoftKeyboard's corpus](https://github.com/AnySoftKeyboard/LanguagePack/blob/master/languages/sardinian/pack/dictionary/LICENSE_for_words_merged.xml.txt), GPL-licensed?

**Comment by UniversalSuperBox** [comment]
I'm not a lawyer, but I am pretty sure that AnySoftKeyboard's corpus is unusable here. The keyboard is licensed under a variety of licenses, most restrictive is GNU LGPL-3. If we were to include a GPLv3 text in the keyboard, the entire keyboard would need to be licensed under GPLv3. I don't think we can manage that.

So an empty corpus, or providing the license for the corpus you have provided, would be preferable.

**Comment by adrmzz** [comment]
Sure, we could go for an empty corpus. But I am not sure that "If we were to include a GPLv3 text in the keyboard, the entire keyboard would need to be licensed under GPLv3." Actually, it appears that [ASK is licensed under an Apache License](https://github.com/AnySoftKeyboard/AnySoftKeyboard), whereas [ASK's Sardinian corpus is under GPL3](https://github.com/AnySoftKeyboard/LanguagePack/blob/master/languages/sardinian/pack/dictionary/LICENSE_for_words_merged.xml.txt), which, in turn, is based in another GPL-d corpus.

**Comment by UniversalSuperBox** [comment]
I'm not sure how AnySoft figured out how to do that, and as far as I'm concerned it's not my business. Licensing a corpus as GPLv3 in the first place made things difficult (since the GPLv3 is a software license, not a license for text or ideas). But I *do* know that GPL and LGPL are only compatible in one direction: You can include LGPL code in a GPL project. You cannot include GPL code in an LGPL project.

I assume it was the dictionary's creator's original intent to never have the dictionary included in anything but a GPL program. Therefore it is incompatible here. Please clear the keyboard's corpus so we can merge this keyboard.

**Comment by adrmzz** [comment]
corpus cleared

File: qml/keys/languages.js [hunk 2]
```diff
@@ -50,6 +50,7 @@ function languageIdToName(languageId)
     if (languageId == "pt")         return i18n.tr("Portuguese");
     if (languageId == "ro")         return i18n.tr("Romanian");
     if (languageId == "ru")         return i18n.tr("Russian");
+    if (languageId == "sc")         return i18n.sc("Sardinian");
```
Comment (UniversalSuperBox) [inline qml/keys/languages.js:53]: This should be `i18n.tr,` not `.sc`. The Sardinian keyboard shows up as a blank spot in the list of keyboards at the moment.

File: plugins/sc/qml/Keyboard_sc.qml [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+            CharKey { label: "p"; shifted: "P"; extended: ["0"]; extendedShifted: ["0"]; rightSide: true; }
+        }
+
+        Row {
+            anchors.horizontalCenter: parent.horizontalCenter;
+            spacing: 0
+
+            CharKey { label: "a"; shifted: "A"; extended: ["à","å","ã","â","á","ä","ª","æ"]; extendedShifted: ["À","Å","Ã","Â","Á","Ä","ª","Æ"]; leftSide: true; }
+            CharKey { label: "s"; shifted: "S"; extended: ["·si","ß","$"]; extendedShifted: ["·SI","ẞ","$"] }
+            CharKey { label: "d"; shifted: "D"; extended: ["·dd"]; extendedShifted: ["·DD"] }}
```
Comment (UniversalSuperBox) [inline plugins/sc/qml/Keyboard_sc.qml:53]: There is an extra closing brace on line 53, which is causing the keyboard to fail to load.

File: plugins/sc/qml/Keyboard_sc.qml [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        Row {
+            anchors.horizontalCenter: parent.horizontalCenter;
+            spacing: 0
+
+            ShiftKey {}
+            CharKey { label: "z"; shifted: "Z"; }
+            CharKey { label: "x"; shifted: "X"; }
+            CharKey { label: "c"; shifted: "C"; extended: ["ç"]; extendedShifted: ["Ç"] }
+            CharKey { label: "v"; shifted: "V"; }
+            CharKey { label: "b"; shifted: "B"; extended: ["·bi","·bos"]; extendedShifted: ["·BI","·BOS"] }}
```
Comment (UniversalSuperBox) [inline plugins/sc/qml/Keyboard_sc.qml:72]: Another double closing brace.

File: plugins/sc/qml/Keyboard_sc_url.qml [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+        Row {
+            anchors.horizontalCenter: parent.horizontalCenter;
+            spacing: 0
+
+            ShiftKey {}
+            CharKey { label: "z"; shifted: "Z"; }
+            CharKey { label: "x"; shifted: "X"; }
+            CharKey { label: "c"; shifted: "C"; extended: ["ç"]; extendedShifted: ["Ç"] }
+            CharKey { label: "v"; shifted: "V"; }
+            CharKey { label: "b"; shifted: "B"; extended: ["·bi","·bos"]; extendedShifted: ["·BI","·BOS"] }}
```
Comment (UniversalSuperBox) [inline plugins/sc/qml/Keyboard_sc_url.qml:72]: Double closing brace

File: plugins/sc/qml/Keyboard_sc_url_search.qml [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+            CharKey { label: "p"; shifted: "P"; extended: ["0"]; extendedShifted: ["0"]; rightSide: true; }
+        }
+
+        Row {
+            anchors.horizontalCenter: parent.horizontalCenter;
+            spacing: 0
+
+            CharKey { label: "a"; shifted: "A"; extended: ["à","å","ã","â","á","ä","ª","æ"]; extendedShifted: ["À","Å","Ã","Â","Á","Ä","ª","Æ"]; leftSide: true; }
+            CharKey { label: "s"; shifted: "S"; extended: ["·si","ß","$"]; extendedShifted: ["·SI","ẞ","$"] }
+            CharKey { label: "d"; shifted: "D"; extended: ["·dd"]; extendedShifted: ["·DD"] }}
```
Comment (UniversalSuperBox) [inline plugins/sc/qml/Keyboard_sc_url_search.qml:53]: Double closing brace

File: plugins/sc/qml/Keyboard_sc_url_search.qml [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+            anchors.horizontalCenter: parent.horizontalCenter;
+            spacing: 0
+
+            ShiftKey {}
+            CharKey { label: "z"; shifted: "Z"; }
+            CharKey { label: "x"; shifted: "X"; }
+            CharKey { label: "c"; shifted: "C"; extended: ["ç"]; extendedShifted: ["Ç"] }
+            CharKey { label: "v"; shifted: "V"; }
+            CharKey { label: "b"; shifted: "B"; extended: ["·bi","·bos"]; extendedShifted: ["·BI","·BOS"] }}
+            CharKey { label: "n"; shifted: "N"; extended: ["·nche","·nde","·nos","ny","ñ"]; extendedShifted: ["·NCHE","·NDE","·NOS","NY","Ñ"] }
```
Comment (UniversalSuperBox) [inline plugins/sc/qml/Keyboard_sc_url_search.qml:73]: Double closing brace

File: plugins/sc/src/src.pro [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+
+# generate database for presage:
+PLUGIN_INSTALL_PATH = $${UBUNTU_KEYBOARD_LIB_DIR}/sc/
+
+lang_db_sc.commands += \
+  rm -f $$PWD/database_sc.db && \
+  text2ngram -n 1 -l -f sqlite -o $$PWD/database_sc.db $$PWD/corpus.txt && \
+  text2ngram -n 2 -l -f sqlite -o $$PWD/database_sc.db $$PWD/corpus.txt && \
+  text2ngram -n 3 -l -f sqlite -o $$PWD/database_sc.db $$PWD/corpus.txt
+lang_db_sc.files += $$PWD/database_sc.db
```
Comment (UniversalSuperBox) [inline plugins/sc/src/src.pro:30]: It appears that the sc database was not rebuilt after removing the corpus: it is still 43MB. Please recreate an empty database as well.

**Review by UniversalSuperBox — CHANGES_REQUESTED** [review]
I tested the keyboard on a device and found that it does not function. Please fix these syntax errors and re-test the keyboard on your device, then let me know you're ready for another review.

**Comment by peat-psuwit** [comment]
Hello,

Thank you for contributing to UBports. As part of project renaming and the effort to port Ubuntu Touch stack to Ubuntu 20.04, we're incrementally migrating repositories to GitLab.

Your PR is now migrated to:
https://gitlab.com/ubports/core/lomiri-keyboard/-/merge_requests/121
However, you might not be able to push new code into it, as the branch resides in our repository. In case you need to update it, please open a new MR from your (new) fork. I don't have a better solution at the moment.

Sorry for your inconvenience.

**Pull request closed without merging**